### ISSUE TITLE
Varnish crashes when defining a director within a director with a sick backend (i.e. with a probe)

### ISSUE THREAD

**Issue by aondio** [varnishcache/varnish-cache#2036]
The following crashes affects VC4.1 and master. The same scenario doesn't crash with 4.0.
## Expected Behaviour

Not sure, but not crashing.
## Current Behaviour

```
Child (19577) Last panic at: Fri, 29 Jul 2016 15:41:10 GMT#012
"Assert error in vdir_add_backend(), vdir.c line 127:#012
 Condition((be) != 0) not true.#012
thread = (cache-main)#012
version = varnish-plus-4.1.3r1 revision abd7781#012
ident = Linux,3.19.0-58-generic,x86_64,-junix,-smalloc,-smalloc,-hcritbit,epoll#012
Backtrace:#012
  0x43af9b: varnishd() [0x43af9b]#012
  0x43b8df: varnishd() [0x43b8df]#012
  0x7f26805c6efb: libvmod_directors.so(vdir_add_backend+0x14b) [0x7f26805c6efb]#012
  0x7f268240b0f5: vgc.so(VGC_function_vcl_init+0xc5) [0x7f268240b0f5]#012
  0x7f268240bbbc: vgc.so(+0x2bbc) [0x7f268240bbbc]#012
  0x446db1: varnishd() [0x446db1]#012
  0x7f26907b9ee1: libvarnish.so(+0x6ee1) [0x7f26907b9ee1]#012
  0x7f26907ba36b: libvarnish.so(+0x736b) [0x7f26907ba36b]#012
  0x7f26907bf908: libvarnish.so(VLU_Fd+0xd8) [0x7f26907bf908]#012
  0x7f26907bb0d3: libvarnish.so(VCLS_Poll+0x133) [0x7f26907bb0d3]#012
```
## Steps to Reproduce (for bugs)

```
varnishtest "Directors inception"

varnish v1 -vcl {
        import directors;

        probe cacheprobe {
                .url = "/";
                .interval = 1s;
                .timeout = 1s;
                .window = 4;
                .threshold = 3;
        }

        backend b01 {
                .host = "host1.com";
                .probe = cacheprobe;
        }

        backend b02 {
                .host = "host2.com";
                .probe = cacheprobe;
        }

        sub vcl_init {
                new hash = directors.hash();
                        hash.add_backend(b01,1.0);
                        hash.add_backend(b02,1.0);

                new rr = directors.round_robin();
                        rr.add_backend(hash.backend("text"));

        }

} -start
```
## Your Environment

- Version used: VC4.1 and master
- Operating System and version: Ubuntu Trusty and EL7.

**Comment by fgsch** [comment]
The crash is due to the backend being sick by default with the definied probe (unless .initial >= .threshold).  In that case hash.backend("text") returns NULL and add_backend() panics.

From a quick glance it looks like it works in < 4.1 by chance. The probe is done after vcl_init{}.
In 4.1 / master, by the time we do the rr.add_backend() the probe has already marked the backend as sick.

**Comment by nigoroll** [comment]
@fgsch this sort-of got assigned to the two us us during bugwash. I'd probably get back to this anyway when I hopefully get back to the vslp->hash merge, but I dunno when. Feel free to work on this if you can

**Comment by Dridi** [comment]
Following today's bugwash I will share my cents here.

We have two different problems in this issue:

- dealing with NULL backends/directors
- having a hash-behind-director setup

With the native directors, all `.backend()` methods return a pointer to their `struct director` for later resolutions. In the case of the hash director (or any director with just-in-time selection) the `.backend()` method takes an argument, and resolves a backend right away.

In this context, using the hash director in `vcl_init` is *useless* and won't work as the original reporter (not @aondio) expected.

Answering to @bsdphk's comment on IRC today:
> the resolve method is maybe not a good idea, it makes it confusing when you have the "real" backend and when it hasn't been decided yet.

This is the mechanism that allows us to stack directors, and for that matter the hash director SHOULD NOT be stacked UNDER another director.

So coming back to the first problem I mention, how do we deal with NULL directors when added to another director?

I don't have a solution but basically if the director cannot defer the backend resolution then it shouldn't be added to another director. Otherwise there's no sensible reason to return NULL.

Can we enforce it (sanely) with libvcc? I don't know.

**Comment by Dridi** [comment]
I made a suggestion that the hash director could have a `.director()` method. The difference with `.backend()` is that it would work like other native directors and return a durable and resolvable backend.

If you use it as an actual backend, instead of working as random-then-sticky it would be just random in the absence of a key at resolve-time (in varnishd, not in the VCL code). But this would mainly allow you to use a hash director with `std.healthy` or stack it under another director for the original reporter's use case of a global health check (several hash directors).

If we do that, should we then fold hash and random directors together? Much like how the client director can be emulated by composing a hash director with `client.identity`.

We discussed this topic today:

> 12:08 < dridi> gquintar: adding an interface was essentially my suggestion for the hash director
12:08 < gquintar> can do
12:08 < dridi> .backend() gives you a backend resolved by the hash director (with a key)
12:09 < dridi> .director() would give a director representation of the hash itself (resolve would yield a random backend)
12:09 < dridi> this way a hash.director() could be stacked under another director for global health probing
12:10 < dridi> opinion? (there's a support ticket where a customer wants to do this kind of health check)
12:10 < dridi> gquintar: I'm asking you because you just went through a thought process touching on this area of varnish
12:10 < gquintar> [off topic]
12:11 < gquintar> for your case, it sounds good
